File: ga/glm4_result_2024_6/result/1968D_cpp_glm/40_1968D.cpp
use std::io::{self, Read};

fn calculate(p: &[usize], a: &[i32], pos: usize, k: usize) -> i64 {
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; p.len()];
    let mut pos = pos;
    let mut k = k;

    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = mx.max(cur + (k as i64) * (a[pos] as i64));
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }

    mx
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut lines = input.lines();

    let t: usize = lines.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let mut line = lines.next().unwrap().split_whitespace();
        let n: usize = line.next().unwrap().parse().unwrap();
        let k: usize = line.next().unwrap().parse().unwrap();
        let ps1: usize = line.next().unwrap().parse().unwrap();
        let ps2: usize = line.next().unwrap().parse().unwrap();

        let p: Vec<usize> = (0..n).map(|_| line.next().unwrap().parse().unwrap() - 1).collect();
        let a: Vec<i32> = (0..n).map(|_| line.next().unwrap().parse().unwrap()).collect();

        let ans1 = calculate(&p, &a, ps1 - 1, k);
        let ans2 = calculate(&p, &a, ps2 - 1, k);

        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}

File: ga/glm4_result_2024_6/result/1968D_glm/40_1968D.rs
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let t = input.trim().parse::<i32>().unwrap();

    for _ in 0..t {
        let mut tokens = input.split_whitespace().map(|s| s.parse::<i32>().unwrap());
        let n = tokens.next().unwrap();
        let k = tokens.next().unwrap();
        let a = tokens.next().unwrap();
        let b = tokens.next().unwrap();

        let mut arr = vec![0; n + 1];
        let mut aa = vec![0; n + 1];

        for x in 1..=n {
            arr[x] = tokens.next().unwrap();
        }

        for x in 1..=n {
            aa[x] = tokens.next().unwrap();
        }

        let mut p1 = 0;
        let mut p2 = 0;
        let mut a = a;
        let mut b = b;
        for x in 1..=k {
            p1 = p1.max(1LL * aa[a] * (k - x + 1));
            p1 += aa[a];
            a = arr[a];

            p2 = p2.max(1LL * aa[b] * (k - x + 1));
            p2 += aa[b];
            b = arr[b];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}